src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpHashJoin.java

```diff
@@ -0,0 +1,79 @@
+package se.liu.ida.hefquin.engine.queryplan.executable.impl.ops;
+
+import org.apache.jena.sparql.core.Var;
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+import se.liu.ida.hefquin.engine.datastructures.SolutionMappingsIndex;
+import se.liu.ida.hefquin.engine.datastructures.impl.SolutionMappingsHashTable;
+import se.liu.ida.hefquin.engine.datastructures.impl.SolutionMappingsHashTableBasedOnOneVar;
+import se.liu.ida.hefquin.engine.datastructures.impl.SolutionMappingsHashTableBasedOnTwoVars;
+import se.liu.ida.hefquin.engine.queryplan.ExpectedVariables;
+import se.liu.ida.hefquin.engine.queryplan.executable.IntermediateResultBlock;
+import se.liu.ida.hefquin.engine.queryplan.executable.IntermediateResultElementSink;
+import se.liu.ida.hefquin.engine.queryproc.ExecutionContext;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+public class ExecOpHashJoin implements BinaryExecutableOp {
+    protected final SolutionMappingsIndex solMHashTableL;
+
+    public ExecOpHashJoin(final ExpectedVariables inputVars1, final ExpectedVariables inputVars2) {
+        final Set<Var> joinVars = new HashSet<>( inputVars1.getCertainVariables());
+        joinVars.retainAll( inputVars2.getCertainVariables() );
+
+        if (joinVars.size() == 1 ){
+            final Var joinVar = joinVars.iterator().next();
+            this.solMHashTableL = new SolutionMappingsHashTableBasedOnOneVar(joinVar);
+        }
+        else if (joinVars.size() == 2){
+            final Iterator<Var> liVar = joinVars.iterator();
+            final Var joinVar1 = liVar.next();
+            final Var joinVar2 = liVar.next();
+
+            this.solMHashTableL = new SolutionMappingsHashTableBasedOnTwoVars(joinVar1, joinVar2);
+        }
+        else{
+            this.solMHashTableL = new SolutionMappingsHashTable(joinVars);
+        }
+    }
+
+    @Override
+    public int preferredInputBlockSize() {
+        //TODO
+        return 10;
```

Comment:
Why 10??

Reply:
To be honest, I am not sure of the input block size of the hash table. 
If I understand properly, the smaller block can produce intermediate results earlier. In this case, perhaps setting block size as 1 makes sense.

Reply:
Yes, I also think that 1 makes more sense here.